pkg/surveyjson/survey.go

```diff
@@ -39,48 +39,75 @@ type JSONSchemaOptions struct {
 	OutErr              io.Writer
 }
 
+type SchemaContext struct {
+	Name                 string
+	Prefixes             []string
+	RequiredFields       []string
+	SchemaType           *JSONSchemaType
+	ParentType           *JSONSchemaType
+	Output               *orderedmap.OrderedMap
+	AdditionalValidators []survey.Validator
+	ExistingValues       map[string]interface{}
+	Definitions          *map[string]*interface{}
+}
+
 // GenerateValues examines the schema in schemaBytes, asks a series of questions using in, out and outErr,
 func (o *JSONSchemaOptions) GenerateValues(schemaBytes []byte, existingValues map[string]interface{}) ([]byte, error) {
 	t := JSONSchemaType{}
 	err := json.Unmarshal(schemaBytes, &t)
 	if err != nil {
 		return nil, errors.Wrapf(err, "unmarshaling schema %s", schemaBytes)
 	}
-	output := orderedmap.New()
-	err = o.recurse("", make([]string, 0), make([]string, 0), &t, nil, output, make([]survey.Validator, 0), existingValues, make(map[string]*interface{}))
+
+	definitions := make(map[string]*interface{})
+
+	ctx := SchemaContext{
+		Name:                 "",
+		Prefixes:             make([]string, 0),
+		RequiredFields:       make([]string, 0),
+		SchemaType:           &t,
+		ParentType:           nil,
+		Output:               orderedmap.New(),
+		AdditionalValidators: make([]survey.Validator, 0),
+		ExistingValues:       existingValues,
+		Definitions:          &definitions,
+	}
+
+	err = o.recurse(ctx)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 	// move the output up a level
-	if root, ok := output.Get(""); ok {
+	if root, ok := ctx.Output.Get(""); ok {
 		bytes, err := json.Marshal(root)
 		if err != nil {
 			return nil, errors.WithStack(err)
 		}
 		return bytes, nil
 	}
-	return make([]byte, 0), fmt.Errorf("unable to find root element in %v", output)
+	return make([]byte, 0), fmt.Errorf("unable to find root element in %v", ctx.Output)
 
 }
 
-func (o *JSONSchemaOptions) handleConditionals(prefixes []string, requiredFields []string, property string, t *JSONSchemaType, parentType *JSONSchemaType, output *orderedmap.OrderedMap, existingValues map[string]interface{}, definitions map[string]*interface{}) error {
-	if parentType != nil {
-		err := o.handleIf(prefixes, requiredFields, property, t, parentType, output, existingValues, definitions)
+func (o *JSONSchemaOptions) handleConditionals(ctx SchemaContext) error {
```

Comment:
JSONSchemaOptions could be also good object to hold that in. If not let's always pass pointer instead of struct itself. Motification on that object will not be applied. 

```suggestion
func (o *JSONSchemaOptions) handleConditionals(ctx *SchemaContext) error {
```

Reply:
I went back and fourth on if I should use a pointer or just copy the value. 

I decided on values as the things that need to be retained are stored as pointers in the struct and the things that do not need to be retained per call are stored by value. 

If we were to pass by a pointer then we would need to start clearing some changes as we come back up the call stack. Things like removing validators and such.

Reply:
Yes. That is why I did not used struct in the first place. It wreck your head sometimes to know if you modifying value or not and it can lead to bugs.

Reply:
It is good change overall but as you have mentioned we need to be careful.